### PR TITLE
[CELEBORN-1122] Metrics supports json format

### PR DESCRIPTION
**Pull request by suizhe007**
### What changes were proposed in this pull request?
If the user does not use prometheus to collect monitoring metrics, but rather some other ones. Using metrics in JSON format would be more user-friendly.The PR supports JSON format for metrics.

### Why are the changes needed?
Ditto.

### Does this PR introduce _any_ user-facing change?
Metrics supports JSON format

### How was this patch tested?
Cluster test.

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+    })
+    absSource.timers().foreach(t => {
+      recordTimer(absSource, t)
+      t.timer.asInstanceOf[CelebornTimer].reservoir
+        .asInstanceOf[ResettableSlidingWindowReservoir].reset()
+    })
+    val metricDatas = new util.ArrayList[MetricData]
+    innerMetrics.synchronized {
+      while (!innerMetrics.isEmpty) {
+        metricDatas.add(innerMetrics.poll)
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:90]: When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        s"${nt.name}_999thPercentile",
+        absSource.reportNanosAsMills(snapshot.get999thPercentile),
+        timestamp,
+        labelNames,
+        labelValues))
+  }
+
+  private def updateInnerMetrics(absSource: AbstractSource, metricData: MetricData): Unit = {
+    innerMetrics.synchronized {
+      if (innerMetrics.size() >= absSource.metricsCapacity) {
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:301]: if condition meet, can we directly return? as remove and add new metricData seems useless

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala (hunk 1 above)
Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:90]: > When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

I think it makes sense , and make some corrections.

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala (hunk 2 above)
Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala]: > if condition meet, can we directly return? as remove and add new metricData seems useless

Thank you for your suggestion. I have resubmitted, please help me review it again.

Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala]: > if condition meet, can we directly return? as remove and add new metricData seems useless

@RexXiong   Thank you for your suggestion. I have resubmitted, please help me review it again.

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala (hunk 1 above)
Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala]: Thank you for your suggestion. I think it makes sense , and make some corrections.

> When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

@RexXiong  Thank you for your suggestion. I think it makes sense , and make some corrections.

File: common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala [hunk 3]
```diff
@@ -125,15 +129,27 @@ class MetricsSystem(
...
-              .newInstance(kv._2, registry, sources.asScala, servletPath)
-            prometheusServlet = Some(servlet.asInstanceOf[PrometheusServlet])
+            if (kv._1 == "jsonServlet") {
+              jsonServlet = Some(servlet.newInstance(
+                kv._2,
+                registry,
+                sources.asScala,
+                jsonServletPath).asInstanceOf[JsonServlet])
+            }
+            if (kv._1 == "prometheusServlet") {
```
[the remark this answers is not in the document]
Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala]: > nit: else if

@RexXiong  Thank you for your suggestion.

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+package org.apache.celeborn.common.metrics.sink
+
+import java.util
+import java.util.Properties
+
+import scala.beans.BeanProperty
+
+import com.codahale.metrics.MetricRegistry
+import io.netty.channel.ChannelHandler.Sharable
+import org.apache.hadoop.shaded.com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:27]: I have no other comments on this pr except this import as it from shaded dependency, WDYT @waitinfuture @pan3793 @FMX @cfmcgrady

Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:27]: > I have no other comments on this pr except this import as it from shaded dependency, WDYT @waitinfuture @pan3793 @FMX @cfmcgrady

@RexXiong   Thank you for your suggestion. I reimport the jackson package,and the version from hadoop 3.2.4.

File: pom.xml [hunk 5]
```diff
@@ -96,6 +96,8 @@
     <zstd-jni.version>1.5.2-1</zstd-jni.version>
     <kubernetes-client.version>6.7.0</kubernetes-client.version>
     <rocksdbjni.version>8.5.3</rocksdbjni.version>
+    <jackson.version>2.10.5</jackson.version>
```
Comment (pan3793) [inline pom.xml:99]: why choose such a low version

**Comment by pan3793** [comment]
please update the LICENSE and NOTICE since it introduces the new dependencies, also, fix sbt building system

File: pom.xml (hunk 5 above)
Comment (suizhe007) [reply pom.xml:99]: > why choose such a low version
The version of `org. apache. hadoop. shaded. *. jackson` comes from Hadoop 3.2.4 now. In order to reduce its impact on the original framework. I directly import the jackson 2.5.1 from Hadoop 3.2.4.

**Comment by suizhe007** [comment]
> please update the LICENSE and NOTICE since it introduces the new dependencies, also, fix sbt building system

@pan3793  Thank you. I have updated newer version , and fixed sbt building system. Please help me review it again.

File: project/CelebornBuild.scala [hunk 6]
```diff
@@ -56,6 +56,8 @@ object Dependencies {
   val ratisVersion = "2.5.1"
   val roaringBitmapVersion = "0.9.32"
   val rocksdbJniVersion = "8.5.3"
+  val jacksonVersion = "2.15.3"
+  val paranamerVersion = "2.8"
```
Comment (pan3793) [inline project/CelebornBuild.scala:60]: seems we don't need to manage `paranamerVersion` explicitly, let's simply delegate it to jackson

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+import com.codahale.metrics.MetricRegistry
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.{ClassTagExtensions, DefaultScalaModule}
+import io.netty.channel.ChannelHandler.Sharable
+
+import org.apache.celeborn.common.metrics.{CelebornHistogram, CelebornTimer, ResettableSlidingWindowReservoir}
+import org.apache.celeborn.common.metrics.source.{AbstractSource, NamedCounter, NamedGauge, NamedHistogram, NamedTimer, Source}
+
+object JsonConverter {
+  val mapper = new ObjectMapper() with ClassTagExtensions
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:33]: I guess user may preview the json using browser, maybe we can print the pretty json string.

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala [hunk 8]
```diff
@@ -49,8 +49,14 @@ class HttpRequestHandler(
     val msg = HttpUtils.handleRequest(service, path, parameters)
     val response = msg match {
       case Invalid.invalid =>
-        if (prometheusHttpRequestHandler != null) {
-          prometheusHttpRequestHandler.handleRequest(uri)
+        if (servletHttpRequestHandlers != null) {
+          val servletHttpRequestHandler = servletHttpRequestHandlers.find(servlet =>
+            uri == servlet.getServletPath())
+          if (servletHttpRequestHandler.nonEmpty) {
+            servletHttpRequestHandler.get.handleRequest(uri)
```
Comment (pan3793) [inline service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:56]: should we update it for JSON?
```
res.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=UTF-8")
```

Comment (suizhe007) [reply service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:56]: > should we update it for JSON?
> 
> ```
> res.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=UTF-8")
> ```

Thank you. You are right. I update it.

File: project/CelebornBuild.scala (hunk 6 above)
Comment (suizhe007) [reply project/CelebornBuild.scala:60]: > seems we don't need to manage `paranamerVersion` explicitly, let's simply delegate it to jackson

Thank you. I remove it.

File: common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala [hunk 9]
```diff
@@ -27,30 +27,34 @@ import scala.util.matching.Regex
...
-    conf: CelebornConf,
-    val servletPath: String) extends Logging {
+    conf: CelebornConf) extends Logging {
   private[this] val metricsConfig = new MetricsConfig(conf)
 
   private val sinks = new ArrayBuffer[Sink]
   private val sources = new CopyOnWriteArrayList[Source]
   private val registry = new MetricRegistry()
+  val prometheusServletPath = conf.get(METRICS_PROMETHEUS_PATH)
+  val jsonServletPath = conf.get(METRICS_JSON_PATH)
```
Comment (cfmcgrady) [inline common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:45]:
```suggestion
  private val prometheusServletPath = conf.get(METRICS_PROMETHEUS_PATH)
  private val jsonServletPath = conf.get(METRICS_JSON_PATH)
```

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+  mapper.registerModule(DefaultScalaModule)
+
+  def toJson(value: ArrayBuffer[Any]): String = {
+    mapper.writeValueAsString(value)
+  }
+
+  def toJsonWithPretty(value: ArrayBuffer[Any]): String = {
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:40]:
```suggestion
  def toPrettyJson(value: Any): String = {
```

File: LICENSE-binary [hunk 11]
```diff
@@ -280,6 +280,10 @@ org.scala-lang:scala-reflect
 org.slf4j:jcl-over-slf4j
 org.yaml:snakeyaml
 org.rocksdb:rocksdbjni
+com.fasterxml.jackson.module:jackson-module-scala
+com.fasterxml.jackson.core:jackson-databind
+com.fasterxml.jackson.core:jackson-annotations
+com.fasterxml.jackson.core:jackson-core
```
Comment (pan3793) [inline LICENSE-binary:286]: LICENSE-binary reflects the final content of the binary release tarball, so the parameter should be retained here.

File: dev/deps/dependencies-client-spark-3.2 [hunk 12]
```diff
@@ -20,10 +20,14 @@ commons-crypto/1.0.0//commons-crypto-1.0.0.jar
 commons-io/2.13.0//commons-io-2.13.0.jar
 commons-lang3/3.12.0//commons-lang3-3.12.0.jar
 commons-logging/1.1.3//commons-logging-1.1.3.jar
-failureaccess/1.0.1//failureaccess-1.0.1.jar
-guava/32.1.3-jre//guava-32.1.3-jre.jar
-hadoop-client-api/3.3.6//hadoop-client-api-3.3.6.jar
-hadoop-client-runtime/3.3.6//hadoop-client-runtime-3.3.6.jar
+guava/14.0.1//guava-14.0.1.jar
+hadoop-client-api/3.2.4//hadoop-client-api-3.2.4.jar
```
Comment (pan3793) [inline dev/deps/dependencies-client-spark-3.2:24]: please rebase master and regenerate this files

File: LICENSE-binary (hunk 11 above)
Comment (suizhe007) [reply LICENSE-binary:286]: > LICENSE-binary reflects the final content of the binary release tarball, so the parameter should be retained here.

what is `the parameter`?

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala (hunk 10 above)
Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala]: Thank you. I update it.

File: common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+import io.netty.channel.ChannelHandler.Sharable
+
+import org.apache.celeborn.common.metrics.{CelebornHistogram, CelebornTimer, ResettableSlidingWindowReservoir}
+import org.apache.celeborn.common.metrics.source.{AbstractSource, NamedCounter, NamedGauge, NamedHistogram, NamedTimer, Source}
+
+object JsonConverter {
+  val mapper = new ObjectMapper() with ClassTagExtensions
+  mapper.registerModule(DefaultScalaModule)
+
+  def toJson(value: ArrayBuffer[Any]): String = {
```
[the remark this answers is not in the document]
Comment (suizhe007) [reply common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala]: Thank you. I update it.

File: dev/deps/dependencies-client-spark-3.2 (hunk 12 above)
Comment (suizhe007) [reply dev/deps/dependencies-client-spark-3.2]: Thank you. I update it.

File: LICENSE-binary (hunk 11 above)
Comment (pan3793) [reply LICENSE-binary:286]: sorry, the paranamer

**Review by RexXiong — APPROVED** [review]
LGTM, cc @pan3793 @cfmcgrady

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala [hunk 14]
```diff
@@ -47,21 +47,34 @@ class HttpRequestHandler(
     val uri = req.uri()
     val (path, parameters) = HttpUtils.parseUri(uri)
     val msg = HttpUtils.handleRequest(service, path, parameters)
-    val response = msg match {
+    val textType = "text/plain; charset=UTF-8"
+    val jsonType = "application/json; charset=UTF-8"
```
Comment (pan3793) [inline service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:51]:
```suggestion
    val jsonType = "application/json"
```

https://www.rfc-editor.org/rfc/rfc8259.html

>    Note:  No "charset" parameter is defined for this registration.
>    Adding one really has no effect on compliant recipients.

**Review by pan3793 — APPROVED** [review]
LGTM, except for one minor thing

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala (hunk 14 above)
Comment (suizhe007) [reply service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:51]: Thanks. I update it.

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala [hunk 15]
```diff
@@ -47,21 +47,34 @@ class HttpRequestHandler(
...
+          val servletHttpRequestHandler = servletHttpRequestHandlers.find(servlet =>
+            uri == servlet.getServletPath())
+          servletHttpRequestHandler match {
+            case Some(handler) => handler match {
+                case jsonHandler: JsonHttpRequestHandler =>
+                  (jsonHandler.handleRequest(uri), jsonType)
+                case _ => (handler.handleRequest(uri), textType)
+              }
+            case _ => (s"Unknown path $uri!", textType)
+          }
```
Comment (cfmcgrady) [inline service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:64]: nit
```suggestion
          servletHttpRequestHandlers.find(servlet => uri == servlet.getServletPath())
            .map {
              case jsonHandler: JsonHttpRequestHandler =>
                (jsonHandler.handleRequest(uri), jsonType)
              case handler: ServletHttpRequestHandler =>
                (handler.handleRequest(uri), textType)
            }.getOrElse((s"Unknown path $uri!", textType))
```

**Review by cfmcgrady — APPROVED** [review]
LGTM, except a nit

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala (hunk 15 above)
Comment (suizhe007) [reply service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:64]: Thanks. I update it.

**Comment by RexXiong** [comment]
Thank you all @suizhe007 @pan3793 @cfmcgrady, merge to main(v0.4.0)

**Pull request closed without merging**